core/trino-main/src/main/java/io/trino/operator/ExchangeClient.java

```diff
@@ -220,49 +203,25 @@ public SerializedPage pollPage()
     {
         assertNotHoldsLock();
 
-        throwIfFailed();
-
         if (closed.get()) {
             return null;
         }
 
-        SerializedPage page = pageBuffer.poll();
+        SerializedPage page = buffer.pollPage();
 
         if (page == null) {
             return null;
         }
 
-        if (page == NO_MORE_PAGES) {
-            // mark client closed; close() will add the end marker
-            close();
-
-            notifyBlockedCallers();
-
-            // don't return end of stream marker
-            return null;
-        }
-
-        synchronized (this) {
-            if (!closed.get()) {
-                bufferRetainedSizeInBytes -= page.getRetainedSizeInBytes();
-                systemMemoryContext.setBytes(bufferRetainedSizeInBytes);
-            }
-            scheduleRequestIfNecessary();
-        }
+        systemMemoryContext.setBytes(buffer.getRetainedSizeInBytes());
+        scheduleRequestIfNecessary();
 
         return page;
     }
 
     public boolean isFinished()
     {
-        throwIfFailed();
-        // For this to works, locations must never be added after is closed is set
-        return isClosed() && completedClients.size() == allClients.size();
-    }
-
-    public boolean isClosed()
-    {
-        return closed.get();
+        return buffer.isFinished() && completedClients.size() == allClients.size();
```

Comment:
why semantics change about `throwIfFailed()` here? The `StreamingExchangeClientBuffer.isFinished` would not throw even if `failure != null`

Reply:
It just seem to be a very weird place to throw in case of an exception and it wasn't necessary. `isFinished` intuitively implies a simple check and one may not expect it to throw if the buffer is failed.